### PR TITLE
probe: handle hostname extension

### PR DESCRIPTION
**Pull request by akshaysngupta**
This PR adds code to handle hostname extension annotation for the health probe host.
Probe host is derived from ingress/rule/host if readiness or liveness probe is not provided.

If there are more then 1 host name, then select the one that doesn't have a `*` or `?`. If none is found, then leave host field with the default value.

This is also responsible for intermittent failures in tests.

### CODE REVIEW

File: pkg/appgw/internaltypes.go [hunk 1]
```diff
@@ -253,3 +257,15 @@ func (listenerID *listenerIdentifier) getHostNames() []string {
 
 	return hostnames
 }
+
+// getFirstHostNameWithouSpecialChars returns the first hostname which doesn't have special chars. To be used for probes.
+func (listenerID *listenerIdentifier) getFirstHostNameWithouSpecialChars() *string {
```
Comment (jaishals) [inline pkg/appgw/internaltypes.go:262]: Nit : Typo

File: pkg/appgw/internaltypes.go [hunk 2]
```diff
@@ -39,6 +40,9 @@ const (
 const (
 	// MaxAllowedHostnames the maximum number of hostnames allowed for listener.
 	MaxAllowedHostnames int = 5
+
+	// WildcardSpecialCharacters are characters that are allowed for wildcard hostnames.
+	WildcardSpecialCharacters = "*?"
```
Comment (jaishals) [inline pkg/appgw/internaltypes.go:45]: Should we validate this in hostname ?

File: pkg/appgw/internaltypes.go (hunk 1 above)
Comment (jaishals) [reply pkg/appgw/internaltypes.go:262]: Shall we call it GetHostNameForProbes

File: pkg/appgw/internaltypes.go (hunk 2 above)
Comment (akshaysngupta) [reply pkg/appgw/internaltypes.go:45]: getHostNames function returns a list with the first hostname being the host in the ingress rule. So this should cover it. Is that what you mean ?

Comment (jaishals) [reply pkg/appgw/internaltypes.go:45]: No I meant, that do we do any validations when user commit the ingress, probably not. I was thinking we should allow users to commit ingress if extendedhostname has any other charactes which we dont support. 
But feel free to ignore as part of this review

Comment (akshaysngupta) [reply pkg/appgw/internaltypes.go:45]: This is what we will need. https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/

**Comment by akshaysngupta** [comment]
This PR adds code to handle hostname extension annotation for the health probe host.
Probe host is derived from ingress/rule/host if readiness or liveness probe is not provided.

If there are more then 1 host name, then select the one that doesn't have a `*` or `?`. If none is found, then leave host field with the default value.

This is also responsible for intermittent failures in tests.

**Pull request merged**